homeassistant/components/mqtt/fan.py

```diff
@@ -217,19 +354,107 @@ def state_received(msg):
                 "qos": self._config[CONF_QOS],
             }
 
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def percentage_received(msg):
+            """Handle new received MQTT message for the percentage."""
+            numeric_val_str = self._templates[ATTR_PERCENTAGE](msg.payload)
+            try:
+                percentage = ranged_value_to_percentage(
+                    self._speed_range, int(numeric_val_str)
+                )
+            except ValueError:
+                _LOGGER.warning(
+                    "'%s' received on topic %s is not a valid speed within the speed range",
+                    msg.payload,
+                    msg.topic,
+                )
+                return
+            if 0 <= percentage <= 100:
+                self._percentage = percentage
+            else:
+                _LOGGER.warning(
+                    "'%s' received on topic %s is not a valid speed within the speed range",
+                    msg.payload,
+                    msg.topic,
+                )
+                return
+            self.async_write_ha_state()
+
+        if self._topic[CONF_PERCENTAGE_STATE_TOPIC] is not None:
+            topics[CONF_PERCENTAGE_STATE_TOPIC] = {
+                "topic": self._topic[CONF_PERCENTAGE_STATE_TOPIC],
+                "msg_callback": percentage_received,
+                "qos": self._config[CONF_QOS],
+            }
+            self._percentage = None
+
+        @callback
+        @log_messages(self.hass, self.entity_id)
+        def preset_mode_received(msg):
+            """Handle new received MQTT message for preset mode."""
+            preset_mode = self._templates[ATTR_PRESET_MODE](msg.payload)
+            if preset_mode in self.preset_modes:
+                self._preset_mode = preset_mode
+                if not self._implemented_percentage and (
+                    preset_mode in self.speed_list
+                ):
+                    self._percentage = ordered_list_item_to_percentage(
+                        self.speed_list, preset_mode
+                    )
+            else:
+                _LOGGER.warning(
+                    "'%s' received on topic %s is not a valid preset mode",
+                    msg.payload,
+                    msg.topic,
+                )
+                return
+
+            self.async_write_ha_state()
```

Comment:
Use a guard clause instead:
```suggestion
            if preset_mode not in self.preset_modes:
                _LOGGER.warning(
                    "'%s' received on topic %s is not a valid preset mode",
                    msg.payload,
                    msg.topic,
                )
                return

            self._preset_mode = preset_mode
            if not self._implemented_percentage and (
                preset_mode in self.speed_list
            ):
                self._percentage = ordered_list_item_to_percentage(
                    self.speed_list, preset_mode
                )
            self.async_write_ha_state()
```